src/main/java/games/diamant/DiamantForwardModel.java

```diff
@@ -0,0 +1,313 @@
+
+            // Initialize game state
+            dgs.nHazardExplosionsOnPath = 0;
+            dgs.nHazardPoissonGasOnPath = 0;
+            dgs.nHazardRockfallsOnPath  = 0;
+            dgs.nHazardScorpionsOnPath  = 0;
+            dgs.nHazardSnakesOnPath     = 0;
+
+            // All the player will participate in next cave
+            for (int p=0; p < dgs.getNPlayers(); p++)
+                dgs.playerInCave.set(p, true);
+
+            // Draw a card and play it
+            DiamantCard card = dgs.mainDeck.draw();
+            dgs.path.add(card);
+            playCard(card, dgs);
+        }
+    }
+
+    /**
+     * Finishes the game and obtains who is the winner
+     * @param dgs: current game state
+     */
+    private void EndGame(DiamantGameState dgs)
+    {
+        int maxGems = 0;
+        List<Integer> bestPlayers = new ArrayList<>();
+
+        for (int p=0; p < dgs.getNPlayers(); p++)
+        {
+            int nGems = dgs.treasureChests.get(p).GetNumberGems();
+            if (nGems > maxGems)
+            {
+                bestPlayers.clear();
+                bestPlayers.add(p);
+                maxGems = nGems;
+            }
+            else if (nGems == maxGems)
+            {
+                bestPlayers.add(p);
+            }
+        }
+
+        boolean moreThanOneWinner = bestPlayers.size() > 1;
+
+        for (int p=0; p < dgs.getNPlayers(); p++)
+        {
+            if (bestPlayers.contains(p)) {
+                if (moreThanOneWinner)
+                    dgs.setPlayerResult(Utils.GameResult.DRAW, p);
+                else
+                    dgs.setPlayerResult(Utils.GameResult.WIN, p);
+            }
+            else
+                dgs.setPlayerResult(Utils.GameResult.LOSE, p);
+        }
+
+        dgs.setGameStatus(Utils.GameResult.GAME_END);
+    }
+
+
+    /**
+     * Gets the possible actions to be played
+     * If the player is not in the cave, only OutOfCave action can be played
+     * If the player is in the cave, there are only two actions: ExitFromCave, ContinueInCave
+     * @param gameState: current game state
+     */
+    @Override
+    protected List<AbstractAction> _computeAvailableActions(AbstractGameState gameState)
+    {
+        DiamantGameState dgs = (DiamantGameState) gameState;
+        ArrayList<AbstractAction> actions = new ArrayList<>();
+
+        // If the player is still in the cave
+        if (dgs.playerInCave.get(gameState.getCurrentPlayer()))
+        {
+            actions.add(new ContinueInCave());
+            actions.add(new ExitFromCave());
+        }
+        else
+            actions.add(new OutOfCave());
+
+        return actions;
+    }
+
+    @Override
+    protected AbstractForwardModel _copy()
+    {
+        return new DiamantForwardModel();
+    }
+
+    /**
+     * Play the card
+     * @param card: card to be played
+     * @param dgs: current game state
+     */
+    private void playCard(DiamantCard card, DiamantGameState dgs)
```

Comment:
All three places that call fm.playCard() have the same two lines immediately preceding:
            DiamantCard card = dgs.mainDeck.draw();
            dgs.path.add(card);
Could these be added into playCard() - removing one parameter, and reducing code duplication?